File: python/meetup/meetup.py
# Import necessary modules from the Python standard library
from calendar import day_name, monthrange, weekday
from datetime import date

# Define a custom exception class for MeetupDayException
class MeetupDayException(Exception):
    pass

# Define a dictionary to map week descriptors to their starting days
START_DAY = {
    'first': 1,
    'second': 8,
    'third': 15,
    'fourth': 22,
    'fifth': 29,
    'teenth': 13,
    'last': None
}

# Define the meetup function
def meetup(year, month, which, week_day):
    # Check if the provided week descriptor is valid
    if which not in START_DAY:
        raise MeetupDayException(f'What is {which}?')

    # Check if the provided week day is valid
    if week_day not in day_name:
        raise MeetupDayException(f'Unknown day: {week_day}')

    # Determine the starting day of the meetup based on the week descriptor
    start_day = START_DAY[which] or monthrange(year, month)[1] - 6

    try:
        # Calculate the weekday (0 = Monday, 1 = Tuesday, etc.) of the starting day
        start_day_wday = weekday(year, month, start_day)
        # Calculate the desired weekday (0 = Monday, 1 = Tuesday, etc.)
        wanted_wday = list(day_name).index(week_day)
        # Calculate the delta between the desired weekday and starting day's weekday
        delta = (wanted_wday - start_day_wday) % 7
        # Calculate the meetup date by adding the delta to the starting day
        return date(year, month, start_day + delta)

    except ValueError:
        # Handle the case where the date is invalid, such as February 30
        raise MeetupDayException("That day does not exist.")

File: javascript/meetup/meetup.js
export const meetup = (year, month, dayOfWeek, descriptor) => {
  // Map of descriptors to the corresponding range of days
  const descriptorMap = {
    'first': 1,
    'second': 8,
    'third': 15,
    'fourth': 22,
    'fifth': 29,
    'last': 35,
    'montheenth': 13,
    'tuesteenth': 19,
    'wednesteenth': 25,
    'thursteenth': 31,
    'friteenth': 26,
    'saturteenth': 23,
    'sunteenth': 30,
  };

  // Map of days of the week to the corresponding numbers
  const dayOfWeekMap = {
    'monday': 1,
    'tuesday': 2,
    'wednesday': 3,
    'thursday': 4,
    'friday': 5,
    'saturday': 6,
    'sunday': 0,
  };

  // Create a new Date object for the first day of the specified month and year
  let date = new Date(year, month - 1, descriptorMap[descriptor]);

  // Use a loop to find the correct day of the week
  while (date.getDay() !== dayOfWeekMap[dayOfWeek]) {
    date.setDate(date.getDate() + 1);
  }

  // Return the date
  return date;
};
